RationaleMCP/0031/differences.md

```diff
@@ -21,50 +21,52 @@ Flat Modelica is designed to avoid such implicit evaluation of parameters, and t
-- Function calls where all input arguments of the function are discrete-time expressions _are discrete-time provided that the callee is pure_.
+### Constant expressions
+In Flat Modelica, a _constant expression_ is more restricted than in full Modelica, by adding the following requirement:
+- Functions called in a constant expression must be `pure constant`.
 
-The rule about when-clause bodies, initial equations, and initial algorithms still holds without exception, so purity of functions doesn't matter there.
+By requiring functions called in a constant expression to be `pure constant`, it is ensured that a constant expression can always be evaluated to a value at translation time.  A function call that is not a constant expression must not be evaluated before simulation starts.
 
-### Continuous-time expressions
+### Parameter expressions
 
-Note that a function call expression where the callee is impure is a non-discrete-time expression, regardless of the variability of the argument expressions.
+In Flat Modelica, a _parameter expression_ is more restricted than in full Modelica, by adding the following requirement:
+- Functions called in a parameter expression must be pure.
+
+As a consequence, the full Modelica syntactic sugar of using an impure function in the binding equation of a parameter is not allowed in Flat Modelica.  Such initialization has to be expressed explicitly using an initial equation.  Hence, the rules of variability hold without exception also in the case of components declared as parameter.
 
 ### Reason for change
 By excluding `external` functions, translation time evaluation of constant expressions is greatly simplified.  By excluding `impure` functions and `pure(…)` expressions, it is ensured that it doesn't matter whether evaluation happens at translation time or at simulation (initialization) time.
 
 Forbidding translation time evaluation of function calls in non-constant expressions generalizes the current Modelica rule for `impure` functions and makes it clear that this is not allowed regardless whether this is seen as an optimization or not.  (The current Modelica specification only has a non-normative paragraph saying that performing optimizations is not allowd.)
 
+The change regarding parameter expressions could be extended to discrete-time expressions as well without loss of expressiveness due to the existing restrictions on where an impure function may be called.  This could also be expressed more generally by saying that a function call expression where the callee is impure is a non-discrete-time expression.  However, it was decided to not include this in the formal description of differences between Modelica and Flat Modelica in order to avoid describing changes that only clarify things without actually making a difference to semantics.
```

Comment:
I don't see how this is true for discrete-time expressions. You are allowed to call impure functions in when-clauses, and the result is a discrete-time expressions. That is sort of the intended way to communicate with external systems.

Reply:
It is true because the rule of capping the variability inside when-clauses to discrete-time still applies.